azure-cli/network/Select-FreeSubnetRange.ps1

```diff
@@ -0,0 +1,202 @@
+function Select-FreeSubnetRange {
+  param(
+    [Parameter(Mandatory = $true)]
+    [int]
+    $SubnetMask,    
+  
+    [Parameter(Mandatory = $true)]
+    [string]
+    $VnetName,
+  
+    [Parameter(Mandatory = $true)]
+    [string]
+    $ResourceGroupName
+  ) 
+  
+  $subnets = az network vnet subnet list `
+    --resource-group $ResourceGroupName `
+    --vnet-name $VnetName `
+    --query "[].addressPrefix"
+    
+  if ($LASTEXITCODE -ne 0) {
+    throw $subnets
+  }
+  
+  Write-Host "Existing subnets:"
+  $subnets = $subnets | ConvertFrom-Json | Sort-Object {
+    $cidr = $_ -Split "/"
+    Write-Host "- $_"
+    ConvertTo-UInt32 -IPAddress $([IPAddress]::Parse($cidr[0]))
+  }
+
+  $requiredAddressCount = [System.Math]::Pow(2, 32 - $SubnetMask)
+  Write-Host "Required address count in new space from input /$SubnetMask is $requiredAddressCount"
+  
+  # Loop through all the existing subnets and try to find a valid CIDR block that can fit between two existing subnets
+  for ($i = 0; $i -lt $subnets.Count - 1; $i++) {
+    # Split the CIDR notation so we have network address and the subnet mask
+    $cidr = $subnets[$i] -split "/"
+
+    # Parse the network address to an IPAddress object and the subnet mask to an integer
+    $currentNetworkAddress = [IPAddress]::Parse($cidr[0])
+    $currentSubnetMask = [int]::Parse($cidr[1])
+
+    # Calculate the broadcast address for this current subnet.
+    # The broadcast address is the highest IP address in the scope.
+    $currentBroadcastAddress = Get-BroadcastIPAddress -IPAddress $currentNetworkAddress -SubnetMask $currentSubnetMask
+
+    # Convert the broadcast address to a unsigned 32 bit integer.
+    # All IPv4 addresses are 32 bit integers. 
+    # Converting the IP addresses to integers allows us to do basic math with the IP addresses and check how far apart they are.
+    $currentBroadcastAddressUInt32 = ConvertTo-UInt32 -IPAddress $currentBroadcastAddress
+
+    # Get the next subnet from the ordered subnet list and also convert it to an integer.
+    $nextNetworkAddress = $($subnets[$i + 1] -split "/")[0]
+    $nextNetworkAddressUInt32 = ConvertTo-UInt32 -IPAddress $nextNetworkAddress
+
+    # Check if there is enough space between the current and the next subnet to fit the new subnet we want to create.
+    if ($nextNetworkAddressUInt32 - $currentBroadcastAddressUInt32 - 1 -ge $requiredAddressCount) {
+      # Add 1 to the broadcast address to get the network IP address of a new subnet
+      $candidateIpAddressUInt32 = [UInt32]($currentBroadcastAddressUInt32 + 1)
+      $candidateIpAddress = ConvertTo-IPAddress -IPAddressAsUInt32 $candidateIpAddressUInt32
+
+      # Ensure that the IP address we found is actually a valid network address.
+      # An example of an invalid network address is 192.168.0.1/23
```

Comment:
Just wanna make sure that you meant to write two distinct "invalid" adresses.. 0.1 and 1.0 ?

Reply:
Nice spotted. This is a mistake, it should be 1.0 :)